web/src/app/services/data-store/data-store.service.ts

```diff
@@ -150,26 +161,43 @@ export class DataStoreService {
       )
       .snapshotChanges()
       .pipe(
+        map(surveyChangeAction),
         map(surveys =>
-          List(
-            surveys
-              .map(a => {
-                const docData = a.payload.doc.data() as DocumentData;
-                const id = a.payload.doc.id;
-                return surveyDocToModel(id, docData);
-              })
-              .filter(DataStoreService.filterAndLogError<Survey>)
-              .filter(
-                survey =>
-                  survey instanceof Survey &&
-                  (survey.generalAccess !== SurveyGeneralAccess.UNLISTED ||
-                    (survey.generalAccess === SurveyGeneralAccess.UNLISTED &&
-                      survey.ownerId === userId))
-              )
-              .map(survey => survey as Survey)
+          surveys.filter(
+            survey =>
+              ![
+                SurveyGeneralAccess.PUBLIC,
+                SurveyGeneralAccess.UNLISTED,
+              ].includes(survey.generalAccess!)
```

Comment:
Filtering must be done on the query itself, not on the result set, since Firestore rule do not allow you to query an item which you do not have access to.

To make this work, you'll need to have two streams, one with changes to public and unlisted, and one with restricted surveys to which the user has been granted access. You'll then need to merge the two.

Reply:
firebase is limited. you can't perform a not-in after an in. you can't use two not-in either. in the restricted case, we have to check if the user is in the acl and if the survey is restricted and if the generalAccess field is not set (for compatibility). there is no way to do this. the only way is to do this in the result set, when the generalAccess field has a value.

regarding the other two... we need two different streams because the users can see the UNLISTED survey only if they are the owner.